blockdevice/encryption/luks/luks.go

```diff
@@ -0,0 +1,121 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package luks provides a way to call LUKS2 cryptsetup.
+package luks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	"github.com/talos-systems/go-blockdevice/blockdevice/encryption"
+	"github.com/talos-systems/go-blockdevice/blockdevice/filesystem/luks"
+	"github.com/talos-systems/go-blockdevice/blockdevice/util"
+)
+
+// LUKS implements LUKS2 encryption provider.
+type LUKS struct {
+	cipher int
+}
+
+const (
+	// AESXTSPlain64Cipher represents aes-xts-plain64 encryption cipher.
+	AESXTSPlain64Cipher = iota
+)
+
+// New creates new LUKS2 encryption provider.
+func New(cipher int) *LUKS {
+	return &LUKS{
+		cipher: cipher,
+	}
+}
+
+// Open runs luksOpen on a device and returns mapped device path.
+func (l *LUKS) Open(deviceName string, key []byte) (string, error) {
+	parts := strings.Split(deviceName, "/")
+	mappedPath := util.PartPathEncrypted(parts[len(parts)-1])
+	parts = strings.Split(mappedPath, "/")
+	mappedName := parts[len(parts)-1]
+
+	err := l.runCommand([]string{"luksOpen", deviceName, mappedName, "--key-file=-"}, key)
+	if err != nil {
+		return "", err
+	}
+
+	return mappedPath, nil
+}
+
+// Encrypt implements encryption.Provider.
+func (l *LUKS) Encrypt(deviceName string, key []byte) error {
+	args := []string{"luksFormat", "--type", "luks2", "--key-file=-"}
```

Comment:
reading man page over and over again, I think this is still a passphrase. it might not be the end of the world, but in that case we need to treat it this way - size can be from X to Y, but it's not fixed (as cryptsetup itself will convert it to proper size given the config)